### ISSUE TITLE
Payment Methods vs Payment Instruments

### ISSUE THREAD

**Issue by tommythorsen** [w3c/webpayments#124]
I feel that section 4.1 (Supported vs Enabled Payment Methods) in the Payment Apps spec proposal has some room for improvement. Right now, a payment method can be "supported" and/or "enabled" or neither. However, it might be the case that the user has configured multiple sets of credentials for a single payment method (two different visa cards for instance).

I would like to propose that we separate the concepts of "supported payment methods" and "enabled payment methods" a bit more. How about the following definitions?

**Before:** "Supported payment method"
**Now:** "Payment Method"
**New definition:** _A Payment Method is a general method or scheme for performing a payment. Examples are "Visa", "MasterCard", "PayPal" etc._

**Before:** "Enabled payment method"
**Now:** "Payment Instrument"
**New definition:** _A Payment Instrument is a specific instance of a Payment Method. Examples are "Visa card ending in 1234", "MasterCard ending in 2345", "PayPal account with username john.doe@example.com"_

With these definitions it makes sense to let a user agent query the payment app for two things:
1. The list of payment methods. The app should have one or more payment methods.
2. The list of payment instruments for the current user or user-agent (zero or more).

It would also be useful if the user-agent could support the following two actions:
1. `configure`: This action prompts the payment app to let the user configure a Payment Instrument (optionally for a given Payment Method). This action should terminate in a call to a new API call `registerPaymentInstrument` upon successful configuration of the instrument.
2. `pay`: Initiates a payment for a given Payment Instrument.

**Comment by adrianhopebailie** [comment]
> I feel that section 4.1 (Supported vs Enabled Payment Methods) in the Payment Apps spec proposal has some room for improvement.

Based on your comment I would agree but only because I think you have misunderstood the intention of payment method.

Our architecture intentionally steers clear of the whole concept of payment instruments. Instead users have payment apps which may simply be a wrapper around a payment instrument or may be something completely different.

A payment method is simply a description of the data that is required in a payment request and the data that is required in the payment response. Implicit in that is the processing of the request that is required to produce an appropriate response.

"Visa" and "MasterCard" are actually poor examples of payment methods as they are really just basic card payments. (i.e. The response should contain a PAN, expiry CVV etc). The reason we have different identifiers for them is to allow merchants to be explicit about WHICH cards they accept because if we just had a payment method called "cards" then that would imply that all merchants accept all cards.

The difference between supported and enabled methods is a difference between the design of the app and the configuration. If I wrote payment app that allowed users to store their credit card details in the app and then send these securely in the payment response I could say that the app _supports_ all of the card payment methods. However, until the user actually stores a card, none of these are _enabled_. 

My app might also _support_ payment via Bitcoin but until the user configures the app to have access to a Bitcoin wallet from which it can send payments that payment method is not _enabled_. To take that further, my app might support 3 different ways of paying by Bitcoin defined as 3 different payment methods. In that case, as soon as my app has access to a Bitcoin wallet to make payments all 3 payment methods become enabled.

**TL;DR:** There is NOT a strong link between a stored payment instrument and an enabled payment method. Storing a payment instrument MAY result in one or more payment methods being enabled but that is entirely dependent on the payment method and how the app is designed.

**Comment by tommythorsen** [comment]
Thanks for a great explanation, Adrian. That makes a lot of sense.

The reason I created this issue is that I'm currently in the process of making a tentative implementation of web payments for the Opera browser. One of my top priorities is to make the payment flow as simple and understandable as possible for the user. With that in mind, consider the following potential flow:
- BACKGROUND: Bob the user has installed the app "PayByCard" which supports payment by all brands of credit card. He has configured two different visa cards -- his work card and his private card.
- STEP 1: Bob clicks "Pay" on a web site which supports payment by visa card.
- STEP 2: The browser (the payment mediator) compares the web site's supported payment methods with the installed payment apps and concludes that PayByCard is able to fulfill the payment request. Bob is shown a dialog where he can choose the PayByCard app (and potentially a whole list of other apps that fulfill the criteria).
- STEP 3: Bob selects PayByCard and is taken to the app where he is given the choice between his two configured visa cards.
- STEP 4: Bob selects his private card, and the payment is ready to be processed.

As a UI developer, I would really like to be able to smash STEP 2 and STEP 3 together so that I could present Bob's two credit cards in the very first screen Bob sees, without him having to go through the extra step of selecting the payment app first. Is there a way the current specification allows for me to do this? Can this be done by having multiple "pay" actions with the same type?

**Comment by adrianhopebailie** [comment]
> As a UI developer, I would really like to be able to smash STEP 2 and STEP 3 together so that I could present Bob's two credit cards in the very first screen Bob sees, without him having to go through the extra step of selecting the payment app first. Is there a way the current specification allows for me to do this? Can this be done by having multiple "pay" actions with the same type?

The short answer is no. You can't have flexibility AND efficiency without some built in intelligence or sensible defaults.

There are a bunch of ways this could be done better:
1. The browser allows the user to pick a default payment app and in this case the user isn't even prompted to pick an app. The browser might be smart enough to see, when a new app is installed, if it offers payment methods that are the same as others enabled by other apps and prompt the user at that time to set a default.
2. The payment app allows the user to set rules for auto-selecting the right payment instrument to use. This is how we imagine payment apps differentiating themselves in the market through great UI.
3. The user installs a different payment app for each payment instrument. If the user likes to pick instruments then this is the easiest way to get the same old experience but I suspect that this paradigm will change over time as this new API allows payments to be made with alternative (better) instruments than cards. 
4. The vision of the interledger protocol is to solve for this matching problem dynamically by routing payments per transaction. So if the merchant and user both supported interledger as a payment method then there would be no need to select an instrument at all. See http://interledger.org for more details.

**Comment by ianbjacobs** [comment]
@tommythorsen,

Thanks for letting us know you are working on implementing the API, and thanks for sharing your thoughts as you do so.

I agree with @adrianhopebailie's observation that we anticipate some of the convenience functionality to be taken up by browsers/mediators and payment apps.

Ian

**Comment by tommythorsen** [comment]
Alright, fair enough. Thanks Adrian and Ian for your answers :)

**Comment by carmellabonomolo** [comment]
Hello,what about a app that bypasses all of the usual nightmare of waiting to clear the long drawn out process just to state that the country you live in needs remitance (which also)there has not been stated in all apps I've been through and that's at least 100,goes directly to the holder of the said "funds" settlements global and overwrites the bennificiary.What's the use of having extreme large amounts of liquid currancy if you can't get to it?
